contracts/rewards/src/state.rs

```diff
@@ -0,0 +1,383 @@
+use std::collections::HashMap;
+
+use cosmwasm_schema::cw_serde;
+use cosmwasm_std::{Addr, Storage, Uint256};
+use cw_storage_plus::{Item, Map};
+
+use crate::{error::ContractError, msg::RewardsParams};
+
+#[cw_serde]
+pub struct Config {
+    pub params: RewardsParams,
+}
+
+#[cw_serde]
+pub struct EpochTally {
+    pub epoch_num: u64,
+    pub contract: Addr,
+    pub total_events: u64,
+    pub participation: HashMap<Addr, u64>,
+    pub rewards_rate: Uint256,
+}
+
+impl EpochTally {
+    #[allow(dead_code)]
+    pub fn new(epoch_num: u64, contract: Addr, rewards_rate: Uint256) -> Self {
+        EpochTally {
+            epoch_num,
+            contract,
+            total_events: 0,
+            participation: HashMap::new(),
+            rewards_rate,
+        }
+    }
+}
+
+#[cw_serde]
+pub struct Event {
+    event_id: String,
+    contract: Addr,
+    epoch_num: u64,
+}
+
+#[cw_serde]
+pub struct Epoch {
+    pub epoch_num: u64,
+    pub block_height_started: u64,
```

Comment:
What is this? The block where the epoch starts at?

Reply:
yes